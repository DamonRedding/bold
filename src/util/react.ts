import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
