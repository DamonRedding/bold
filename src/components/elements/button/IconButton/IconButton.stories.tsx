import { boolean, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes } from '../../../../stories-addons/withPropTypes'
import { withTheme } from '../../../../stories-addons/withTheme'

import { IconButton } from './IconButton'

storiesOf('Components', module)
    .addDecorator(withPropTypes(``))
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .add('IconButton', () => (
        <IconButton
            icon='pen'
            disabled={boolean('disabled', false)}
            loading={boolean('loading', false)}
        />
    ))
