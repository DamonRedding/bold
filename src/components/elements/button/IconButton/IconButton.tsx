import * as React from 'react'

import { focusBoxShadow, Theme, withStyles, WithStylesProps } from '../../../../styles'
import { BaseButton, BaseButtonProps } from '../../button/BaseButton'
import { Icons } from '../../Icon/generated/Icons'
import { Icon } from '../../Icon/Icon'

export interface IconButtonProps extends BaseButtonProps, WithStylesProps {
    icon: Icons
    loading?: boolean
}

export interface IconButtonState {
    loading: boolean
}

export const createStyles = (theme: Theme) => ({
    button: {
        backgroundColor: 'transparent',
        border: 'none',
        borderRadius: theme.baseRadius + 2,
        color: theme.color.gray40,
        cursor: 'pointer',
        display: 'inline-flex',
        padding: 0,
        position: 'relative',
        userSelect: 'none',
        transition: 'all .2s',
        transform: 'translate3d(0,0,0)',
        'span': {
            display: 'inline-flex',
            transition: 'color .2s',
        },
        ':disabled': {
            opacity: 0.5,
        },
        ':focus': {
            outline: 'none',
            boxShadow: focusBoxShadow(theme, 'primary', 'single'),
        },
        ':hover': {

        },
    },
    loading: {
        pointerEvents: 'none',
        'span': {
            visibility: 'hidden',
        },
        ':after': {
            animation: `${theme.animation.spinAround} 500ms infinite linear`,
            border: '3px solid currentColor',
            borderRadius: '50%',
            borderRightColor: 'transparent',
            borderTopColor: 'transparent',
            content: '""',
            display: 'block',
            fontSize: '0.875rem',
            height: '1.5em',
            width: '1.5em',
            position: 'absolute',
            left: 'calc(50% - (1.5em / 2))',
            top: 'calc(50% - (1.5em / 2))',

        },
    },
})

@withStyles
export class IconButton extends React.Component<IconButtonProps, IconButtonState> {

    constructor(props, context?) {
        super(props, context)

        this.state = {
            loading: false,
        }
    }

    render() {
        const {
            css,
            icon,
            loading,
            theme,
            ...rest,
        } = this.props

        const styles = createStyles(theme)

        const classes = css(
            styles.button,
            (this.state.loading || loading) && styles.loading
        )

        return (
            <BaseButton
                {...rest}
                className={classes}
                onLoadingChange={this.onLoadingChange}
            >
                <span>
                    <Icon icon={icon} />
                </span>
            </BaseButton>
        )
    }

    private onLoadingChange = (loading: boolean) => {
        this.setState({ loading })
    }

}
