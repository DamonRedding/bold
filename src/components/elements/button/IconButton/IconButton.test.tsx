import { mount, render, shallow } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../../test'

import { IconButton } from './IconButton'

describe('IconButton', () => {
    it('deve renderizar corretamente', () => {
        const wrapper = render(withTheme(<IconButton icon='lapis' />))
        expect(wrapper).toMatchSnapshot()
    })

    it('deve ter animação de "carregando" ao especificar onClick com retorno do tipo Promise', () => {
        const delayedFunction = () => {
            return new Promise((resolve, reject) => {
                setTimeout(resolve, 10)
            })
        }

        const wrapper = mount(withTheme(<IconButton icon='lapis' onClick={delayedFunction} />))
        expect(wrapper).toMatchSnapshot()

        wrapper.simulate('click')
        expect(wrapper).toMatchSnapshot()
    })

    it('não deve ter animação com "onClick" que não seja promise', () => {
        const func = () => undefined
        const wrapper = mount(withTheme(<IconButton icon='lapis' onClick={func} />))
        expect(wrapper).toMatchSnapshot()

        wrapper.simulate('click')
        expect(wrapper).toMatchSnapshot()
    })

    it('deve adicionar um hint string automaticamente', () => {
        const wrapper = render(withTheme(<IconButton icon='lapis' hint='hint teste' />))
        expect(wrapper).toMatchSnapshot()
    })

    it('deve adicionar um hint custom automaticamente', () => {
        const wrapper = render(withTheme(<IconButton icon='lapis' hint={(<span>hint complexo teste</span>)} />))
        expect(wrapper).toMatchSnapshot()
    })

    it('deve ter animação de loading', () => {
        const wrapper = render(withTheme(<IconButton icon='lapis' loading={true} />))
        expect(wrapper).toMatchSnapshot()
    })
})
