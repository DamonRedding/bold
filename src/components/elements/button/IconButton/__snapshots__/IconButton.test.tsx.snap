// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton deve adicionar um hint custom automaticamente 1`] = `
.emotion-2 {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<span
  class="hint-wrapper"
>
  <button
    class="emotion-2"
    type="button"
  >
    <span>
      <span
        class="emotion-1"
      >
        <svg
          aria-hidden="true"
          class="emotion-0"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <g
            id="Editar"
          >
            <path
              d="M4 17.18L4.01 20h2.83L18.01 8.83 15.18 6 4 17.18zM19.72 5.61l-1.33-1.33a.94.94 0 0 0-1.33 0L16 5.33 18.67 8l1-1.06a.94.94 0 0 0 .05-1.33z"
            />
          </g>
        </svg>
      </span>
    </span>
  </button>
</span>
`;

exports[`IconButton deve adicionar um hint string automaticamente 1`] = `
.emotion-2 {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<span
  class="hint-wrapper"
>
  <button
    class="emotion-2"
    type="button"
  >
    <span>
      <span
        class="emotion-1"
      >
        <svg
          aria-hidden="true"
          class="emotion-0"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
        >
          <g
            id="Editar"
          >
            <path
              d="M4 17.18L4.01 20h2.83L18.01 8.83 15.18 6 4 17.18zM19.72 5.61l-1.33-1.33a.94.94 0 0 0-1.33 0L16 5.33 18.67 8l1-1.06a.94.94 0 0 0 .05-1.33z"
            />
          </g>
        </svg>
      </span>
    </span>
  </button>
</span>
`;

exports[`IconButton deve renderizar corretamente 1`] = `
.emotion-2 {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<button
  class="emotion-2"
  type="button"
>
  <span>
    <span
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        class="emotion-0"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <g
          id="Editar"
        >
          <path
            d="M4 17.18L4.01 20h2.83L18.01 8.83 15.18 6 4 17.18zM19.72 5.61l-1.33-1.33a.94.94 0 0 0-1.33 0L16 5.33 18.67 8l1-1.06a.94.94 0 0 0 .05-1.33z"
          />
        </g>
      </svg>
    </span>
  </span>
</button>
`;

exports[`IconButton deve ter animação de loading 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

.emotion-2 {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  pointer-events: none;
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-2 span {
  visibility: hidden;
}

.emotion-2:after {
  -webkit-animation: animation-1scjtrl 500ms infinite linear;
  animation: animation-1scjtrl 500ms infinite linear;
  border: 3px solid currentColor;
  border-radius: 50%;
  border-right-color: transparent;
  border-top-color: transparent;
  content: "";
  display: block;
  font-size: 0.875rem;
  height: 1.5em;
  width: 1.5em;
  position: absolute;
  left: calc(50% - (1.5em / 2));
  top: calc(50% - (1.5em / 2));
}

<button
  class="emotion-2"
  type="button"
>
  <span>
    <span
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        class="emotion-0"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <g
          id="Editar"
        >
          <path
            d="M4 17.18L4.01 20h2.83L18.01 8.83 15.18 6 4 17.18zM19.72 5.61l-1.33-1.33a.94.94 0 0 0-1.33 0L16 5.33 18.67 8l1-1.06a.94.94 0 0 0 .05-1.33z"
          />
        </g>
      </svg>
    </span>
  </span>
</button>
`;
